lavalink/NodeManager.py

```diff
@@ -0,0 +1,210 @@
+
+    @classmethod
+    def third_two(cls):
+        """ Asia and Oceania """
+        return cls(["hongkong", "japan", "singapore", "sydney"])
+
+    @classmethod
+    def third_three(cls):
+        """ North and South America """
+        return cls(["us-central", "us-east", "us-south", "us-west", "vip-us-east", "vip-us-west"])
+
+
+class LavalinkNode:
+    def __init__(self, manager, host, password, regions, rest_port: int = 2333, ws_port: int = 80,
+                 ws_retry: int = 10, shard_count: int = 1):
+        self.regions = regions
+        self._lavalink = manager._lavalink
+        self.manager = manager
+        self.rest_uri = 'http://{}:{}/loadtracks?identifier='.format(host, rest_port)
+        self.password = password
+
+        self.ws = WebSocket(
+            manager._lavalink, self, host, password, ws_port, ws_retry, shard_count
+        )
+        self.server_version = 2
+        self.stats = Stats()
+
+        self.ready = asyncio.Event(loop=self._lavalink.loop)
+
+        self.players = PlayerManager(self._lavalink, self, DefaultPlayer)
+
+    def set_online(self):
+        self.manager.on_node_ready(self)
+
+    def set_offline(self):
+        self.manager.on_node_disabled(self)
+
+    async def manage_failover(self):
+        if self.manager.nodes:
+            new_node = self.manager.nodes[0]
+            for g in list(self.players._players):
+                new_player = self.players._players.pop(g)
+                new_player.node = new_node
+                log.info(new_player.is_playing)
+                is_playing = bool(new_player.is_playing)
+                current_track = copy.copy(new_player.previous)
+                current_posit = copy.copy(new_player._prev_position)
+                new_node.players._players.update({g: new_player})
+                ws = self._lavalink.bot._connection._get_websocket(int(g))
+                await ws.voice_state(int(g), None)
+                if is_playing is True:
+                    await new_player.connect(new_player.channel_id)
+                    new_player.queue.insert(0, current_track)
+                    await new_player.play()
+                    await new_player.seek(current_posit)
+
+
+class NodeManager:
+    def __init__(self, lavalink, default_node, round_robin):
+        self._lavalink = lavalink  # lavalink client
+
+        self.default_node_index = default_node  # index of the default node for REST
+        self.round_robin = round_robin  # enable round robin load balancing
+        self._rr_pos = 0  # starting sound robin position
+
+        self.nodes = []  # list of nodes (online)
+        self.offline_nodes = []  # list of nodes (offline or not set-up yet)
+        self.nodes_by_region = {}  # dictionary of nodes with region keys
+
+        self.ready = asyncio.Event(loop=self._lavalink.loop)
+
+    def __iter__(self):
+        for node in self.nodes:
+            yield node
+
+    def on_node_ready(self, node):
```

Comment:
Perhaps we could have node events that are dispatched (maybe through hooks other than `track_hooks`) so the user can do stuff accordingly.